library/src/contraction/contraction_solution_impl.hpp

```diff
@@ -232,10 +389,723 @@ namespace hiptensor
+            }
+
+            // CK has its own format for indices...
+            auto toCKVec = [](std::vector<std::size_t> const& v) {
+                return std::vector<ck::index_t>(v.begin(), v.end());
+            };
+
+            //Allocate Real and Imaginary inputs
+            auto elementsA  = elementSpaceFromLengthsAndStrides(a_ms_ks_lengths, a_ms_ks_strides);
+            auto elementsB  = elementSpaceFromLengthsAndStrides(b_ns_ks_lengths, b_ns_ks_strides);
+            auto elementsD  = elementSpaceFromLengthsAndStrides(ds_ms_ns_lengths, ds_ms_ns_strides);
+            elementsE       = elementSpaceFromLengthsAndStrides(e_ms_ns_lengths, e_ms_ns_strides);
+
+            A_d_real.reset(nullptr);
+            A_d_real = std::move(allocDevice<elementType>(elementsA));
+            A_d_imag.reset(nullptr);
+            A_d_imag = std::move(allocDevice<elementType>(elementsA));
+            B_d_real.reset(nullptr);
+            B_d_real = std::move(allocDevice<elementType>(elementsB));
+            B_d_imag.reset(nullptr);
+            B_d_imag = std::move(allocDevice<elementType>(elementsB));
+            D_d_real.reset(nullptr);
+            D_d_real = std::move(allocDevice<elementType>(elementsD));
+            D_d_imag.reset(nullptr);
+            D_d_imag = std::move(allocDevice<elementType>(elementsD));
+            E_d_real.reset(nullptr);
+            E_d_real = std::move(allocDevice<elementType>(elementsE));
+            E_d_imag.reset(nullptr);
+            E_d_imag = std::move(allocDevice<elementType>(elementsE));
+
+            auto blockDim = dim3(1024, 1, 1);
+            auto gridDim  = dim3(ceilDiv(elementsA, blockDim.x), 1, 1);
+            hipLaunchKernelGGL(
+                            (unpack<DataType, elementType>), gridDim, blockDim, 0, 0, (const DataType*)A,
+                            (elementType*)A_d_real.get(), (elementType*)A_d_imag.get(), elementsA);
+
+            gridDim  = dim3(ceilDiv(elementsB, blockDim.x), 1, 1);
+            hipLaunchKernelGGL(
+                            (unpack<DataType, elementType>), gridDim, blockDim, 0, 0, (const DataType*)B,
+                            (elementType*)B_d_real.get(), (elementType*)B_d_imag.get(), elementsB);
+
+            gridDim  = dim3(ceilDiv(elementsD, blockDim.x), 1, 1);
+            hipLaunchKernelGGL(
+                            (unpack<DataType, elementType>), gridDim, blockDim, 0, 0, (const DataType*)D,
+                            (elementType*)D_d_real.get(), (elementType*)D_d_imag.get(), elementsD);
+
+            gridDim  = dim3(ceilDiv(elementsE, blockDim.x), 1, 1);
+            hipLaunchKernelGGL(
+                            (unpack<DataType, elementType>), gridDim, blockDim, 0, 0, (const DataType*)E,
+                            (elementType*)E_d_real.get(), (elementType*)E_d_imag.get(), elementsE);
+
+            // Initialize the argument pointer
+            mArgPtr[0] = std::move(deviceOp->MakeArgumentPointer(
```

Comment:
You might want to create a lambda helper function to create each argument pointer so reduce some of the repetition here. The input args to the lambda would be the items that change (device ptrs, etc). The rest would be captures (such as lengths / strides)